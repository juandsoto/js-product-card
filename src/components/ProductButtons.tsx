import React, { useContext, CSSProperties, useCallback } from 'react';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
  const { counter, increaseBy, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(() => {
    return !!maxCount && counter === maxCount;
  }, [counter, maxCount]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        disabled={isMaxReached()}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  );
};
