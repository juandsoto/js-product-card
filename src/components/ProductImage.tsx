import React, { useContext, CSSProperties } from 'react';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img = '', className, style }: Props) => {
  const { product } = useContext(ProductContext);
  let imageToUse: string = '';
  if (img) {
    imageToUse = img;
  } else if (product.img) {
    imageToUse = product.img;
  } else {
    imageToUse = noImage;
  }

  return (
    <img
      style={style}
      className={`${styles.productImg} ${className}`}
      src={imageToUse}
      alt="Coffe Mug"
    />
  );
};
