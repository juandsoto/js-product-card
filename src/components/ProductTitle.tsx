import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
  title?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: Props) => {
  const { product } = useContext(ProductContext);

  let titleToUse: string = title ? title : product.title;

  return (
    <span style={style} className={`${styles.productDescription} ${className}`}>
      {titleToUse}
    </span>
  );
};
