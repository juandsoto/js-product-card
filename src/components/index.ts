import { ProductCardHocProps } from '../interfaces/interfaces';

import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';

export type { Props as ProductCardProps } from './ProductCard';
export type { Props as ProductTitleProps } from './ProductTitle';
export type { Props as ProductImageProps } from './ProductImage';
export type { Props as ProductButtonsProps } from './ProductButtons';

export const ProductCard: ProductCardHocProps = Object.assign(ProductCardHOC, {
	Title: ProductTitle,
	Image: ProductImage,
	Buttons: ProductButtons
});

export default ProductCard;
