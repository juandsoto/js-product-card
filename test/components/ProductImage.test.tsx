import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 as product } from '../data/products';

describe('ProductImage', () => {
  test('debe de mostrar el componente correctamente con la imagen personalizada', () => {
    const wrapper = renderer.create(
      <ProductImage className="custom-image" img={product.img} />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe de mostrar el componente con la imagen por defecto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product}>
        {() => {
          return <ProductImage className="custom-image" />;
        }}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
