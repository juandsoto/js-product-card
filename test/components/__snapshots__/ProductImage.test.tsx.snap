// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostrar el componente con la imagen por defecto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffe Mug"
    className="productImg custom-image"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage debe de mostrar el componente correctamente con la imagen personalizada 1`] = `
<img
  alt="Coffe Mug"
  className="productImg custom-image"
  src="./coffee-mug.png"
/>
`;
